### PR TITLE
SALTO-3318: Salesforce indicative fetch warnings for omitted data instances without name

### PR DESCRIPTION
**Pull request by tamtamirr**
Salesforce indicative fetch warnings for omitted data instances without name.

---

In case none of the `saltoIdFields` has a value, the name of the instance will be empty, which is resolved to instance with the name `_config`. This caused a non indicative fetch warning that specified that those instances were removed due to duplication of `elemID` `_config` which is not the actual underlying issue. 

This was the collision warning `before`:

```
Omitted 36 instances of SBQQ__CustomAction__c due to Salto ID collisions.
Current Salto ID configuration for SBQQ__CustomAction__c is defined as [EmptyField__c].

Breakdown per colliding Salto ID:
- _config:
	* https://salto68-dev-ed.develop.my.salesforce.com/a0B8d000008r49QEAQ
	* https://salto68-dev-ed.develop.my.salesforce.com/a0B8d000008r49REAQ
	* https://salto68-dev-ed.develop.my.salesforce.com/a0B8d000008r49SEAQ
	* 33 more instances

To resolve these collisions please take one of the following actions and fetch again:
	1. Change SBQQ__CustomAction__c's saltoIDSettings to include all fields that uniquely identify the type's instances.
	2. Delete duplicate instances from your salesforce account.

Alternatively, you can exclude SBQQ__CustomAction__c from the data management configuration in salesforce.nacl

Learn more at: https://help.salto.io/en/articles/6927217-salto-for-salesforce-cpq-support
```

This is the collision warning `after` my PR:

```
Omitted 36 instances of SBQQ__CustomAction__c due to Salto ID collisions.
Current Salto ID configuration for SBQQ__CustomAction__c is defined as [EmptyField__c].

Breakdown per colliding Salto ID:
- Instances with empty name (Due to no values in any of the provided ID fields):
	* https://salto68-dev-ed.develop.my.salesforce.com/a0B8d000008r49QEAQ
	* https://salto68-dev-ed.develop.my.salesforce.com/a0B8d000008r49REAQ
	* https://salto68-dev-ed.develop.my.salesforce.com/a0B8d000008r49SEAQ
	* 33 more instances

To resolve these collisions please take one of the following actions and fetch again:
	1. Change SBQQ__CustomAction__c's saltoIDSettings to include all fields that uniquely identify the type's instances.
	2. Delete duplicate instances from your salesforce account.

Alternatively, you can exclude SBQQ__CustomAction__c from the data management configuration in salesforce.nacl

Learn more at: https://help.salto.io/en/articles/6927217-salto-for-salesforce-cpq-support
```

---
_Release Notes_: 
Salesforce Adapter:
- More indicative fetch warning when data instances are omitted due to empty names.

---
_User Notifications_: 
_None_

### CODE REVIEW

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts [hunk 1]
```diff
@@ -126,8 +127,23 @@ const createWarnings = async (
...
   Types of the omitted instances are: ${typesOfIllegalRefSources.join(', ')}.`)]
 
+  const instancesWithoutNameByType = await groupByAsync(
+    instancesWithoutName,
+    async instance => apiName(await instance.getType())
+  )
+
+  const instancesWithoutNameWarnings = Object.entries(instancesWithoutNameByType)
+    .map(([typeName, instances]) => {
+      const message = `Omitted instances without names of type ${typeName}.
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:137]: @tomermevorach Your thought on the fetch warning message?
Tried to make is as similar as possible to [this](https://app.datadoghq.com/logs?query=Omitted%20_config&cols=host%2Cservice&context_event=AYYiJeOqAAAMrQMDp0_HQwLy&event=AgAAAYYiDN1D8xJcDwAAAAAAAAAYAAAAAEFZWWlETjZXQUFDN1cxeU13c1lvLUFLTgAAACQAAAAAMDE4NjIyMGQtMTAyYS00ODY3LTgyYmMtMjg5NTk4ZWQ4OGVi&index=&messageDisplay=inline&stream_sort=time%2Cdesc&viz=&from_ts=1674314665657&to_ts=1675610665657&live=true) warning messages.

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts [hunk 2]
```diff
@@ -272,7 +288,11 @@ const filter: RemoteFilterCreator = ({ client, config }) => ({
       internalToInstance,
       dataManagement,
     )
-    const instancesWithCollidingElemID = getInstancesWithCollidingElemID(customObjectInstances)
+    const instancesWithoutName = customObjectInstances
+      .filter(instance => instance.elemID.name === ElemID.CONFIG_NAME)
```
Comment (adamro) [inline packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:292]: Are you sure that instances without name at this stage actually have `ElemID.CONFIG_NAME` as their `instance.elemID.name` and not an empty string?
Did you test this manually? (ie. change the config to be a field that you have instances with empty value in it and see what happens at this stage)

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts (hunk 1 above)
Comment (tomermevorach) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:137]: lgtm

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts (hunk 2 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:291]: @adamro Yes :) Upated the PR description to include the actual fetch warning as-well

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts (hunk 1 above)
Comment (adamro) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts]: You are missing the 
`Current Salto ID configuration for ${type} is defined as [${getIdFieldsByType(type).join(', ')}].`
which indicates which fields are defined as idFields now, which is important to the user in this case.

Also for all other collision cases we don't actually log the whole instances. We put a link to the service url.

You completely changed the format and the information we show the user just case the name is empty. I don't understand why.

In my opinion you should remove this too special handling and add a case inside `getInstanceDesc` - [here](https://github.com/salto-io/salto/blob/a11f9d4671de9288425fdd22a39b5d84f3a45626/packages/adapter-utils/src/collisions.ts#L41) 
when the name is `ElemID.CONFIG_NAME` to have a more user friendly indication the values are empty.

Comment (tamtamirr) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts]: @adamro My bad. I didn't fully understand the displayed message before, instead I referred to a non user facing log I saw in `datadog`. Decided to move my logic into the `collisions` utility of `adapter-utils`. FYI @netama @alonstern can you take a quick look at this one? I've updated the PR description with the new message that will be be displayed.

Also, I wanna make sure I don't cause any problem when we are actually dealing with `_config` instances. Did I possibly overshadow a correct collision warning there? Are there any cases where the previous warning was actually correct?

**Review by adamro — APPROVED** [review]
🔥

**Pull request merged**